### DISCUSSION TITLE
Memory Leak in nodejs version 20.16.0

### DISCUSSION THREAD

**Discussion by SteffenHo** [nodejs/help#4634] [Q&A] [answered]
We saw a massive increase in memory in our next.js application after updating to version 20.16.0. We build our application with docker and the 20-alpine image.  Other users were also able to confirm this problem: https://www.reddit.com/r/node/comments/1ejzn64/sudden_inexplicable_memory_leak_on_new_builds/ 

![image](https://github.com/user-attachments/assets/ec727423-d85b-488e-946d-b42470383d7b)

We then switched to version 20.15.1 and the memory returned to normal.

Maybe someone else could confirm this problem too?

**Comment by avivkeller** [comment]
Hi! Could you possibly narrow down where the error is coming from and provide a minimal reproducible example?

Additionally, does this also occur on v22?

**Reply by SteffenHo** [reply]
I'm sorry. I could not provide a minimal reproducible example. We have an enterprise shop application with several hundred/thousands pages renderings per minute. We were only able to reproduce the error under higher load.

I therefore opened a discussion and not a bug to find out whether more developers could observe a similar behavior.

**Reply by khuezy** [reply]
Thanks for bringing this up. Here's a related issue with a min repro: https://github.com/vercel/next.js/discussions/68636

**Comment by snyamathi** [comment]
@RedYetiDev I was able to reproduce the issue locally and trace it back to this commit: https://github.com/nodejs/node/commit/45c5ec82a84cce29fbaab5ea10e7d00b9ebf2664

the prior commit, e28292572e837ab79acc4d839a9c044afb77ef30 works with no memory leak

the next commit, 45c5ec82a84cce29fbaab5ea10e7d00b9ebf2664 causes a memory leak

-----

Something was introduced in undici between 6.15.0 and 6.16.1 which is causing issues

**Reply by avivkeller** [reply]
@nodejs/undici

**Reply by snyamathi** [reply]
I'm able to trace it back to this commit https://github.com/nodejs/undici/commit/63b7794af7c282d543a56d13521f9d5e92fa61ae

pretty sure it has something to do with https://github.com/nodejs/undici/blob/abec615bef89992e9fec5dd35c8f59aa33e00459/lib/web/fetch/response.js#L529

**Reply by snyamathi** [reply]
@RedYetiDev https://github.com/nodejs/undici/pull/3445 This fixes the issue locally

**Comment by manuelseisl** [comment]
same issue, when using next.js application with node 20.16.0 on server we saw a massive increase in memory. upgrade to node 22.7.0 solved the issue

**Comment by mcollina** [comment, marked as answer]
This is solver in Node.js v22.7.0. The fix will be available in the next v20 release too.

**Reply by khuezy** [reply]
A heads up, 22.7.0 has a critical bug w/ buffers: https://github.com/nodejs/node/issues/54518
Your app may crash if it uses buffers.
